synapse/rest/client/keys.py

```diff
@@ -109,9 +114,15 @@ async def on_POST(
                 400, "To upload keys, you must pass device_id when authenticating"
             )
 
-        result = await self.e2e_keys_handler.upload_keys_for_user(
-            user_id, device_id, body
-        )
+        if self._is_master:
+            result = await self.e2e_keys_handler.upload_keys_for_user(
+                user_id, device_id, body
+            )
+        else:
+            result = await self.key_uploader(
+                user_id=user_id, device_id=device_id, body=body
+            )
```

Comment:
It can be hard to coerce mypy to like it, but if we can combine these calls it is usually nicer; e.g. in `__init__` something like:

```python
if is_main:
    self.key_uploader = self.e2e_keys_handler.upload_keys_for_user
else:
    self.key_uploader = ReplicationUploadKeysForUserRestServlet.make_client(hs)
```

Then we can just call it as:

```python
result = await self.key_uploader(user_id=user_id, device_id=device_id, body=body)
```

Reply:
I can just call the equivalent of 'is_master' there, right?

Reply:
> I can just call the equivalent of 'is_master' there, right?

`is_main = hs.config.worker.work_app is None`

(which you could just inline)

Reply:
(I wish they'd put the resolve button on the other side on mobile). Why isn't that a helper function anyways? I think I did that pretty much exact thing above somewhere. So just move it into the key_handler instead.

Reply:
> I feel like that's a much larger project than this.

Yes, definitely! I'd submit what you currently have. Maybe file a follow-up issue about reducing all this duplication.